src/main/java/rx/internal/operators/BufferUntilSubscriberV2.java

```diff
@@ -0,0 +1,348 @@
+            if (n < 0L) {
+                throw new IllegalArgumentException("n >= 0 required");
+            } else
+            if (n > 0L) {
+                BackpressureUtils.getAndAddRequest(this, n);
+                replay();
+            } else
+            if (done) { // terminal events can be delivered for zero requests
+                replay();
+            }
+        }
+        /**
+         * Tries to set the given subscriber if not already set, sending an
+         * IllegalStateException to the subscriber otherwise.
+         * @param subscriber
+         */
+        public void setSubscriber(Subscriber<? super T> subscriber) {
+            if (this.subscriber.compareAndSet(null, subscriber)) {
+                subscriber.add(Subscriptions.create(this));
+                subscriber.setProducer(this);
+            } else {
+                subscriber.onError(new IllegalStateException("Only a single subscriber is allowed"));
+            }
+        }
+        /**
+         * Tries to replay the contents of the queue.
+         */
+        void replay() {
+            synchronized (this) {
+                if (emitting) {
+                    missed = true;
+                    return;
+                }
+                emitting = true;
+            }
+            Queue<Object> q = queue;
+            for (;;) {
+                Subscriber<? super T> s = subscriber.get();
+                boolean unlimited = false;
+                if (s != null) {
+                    boolean d = done;
+                    boolean empty = q.isEmpty();
+                    
+                    if (checkTerminated(d, empty, s)) {
+                        return;
+                    }
+                    long r = get();
+                    unlimited = r == Long.MAX_VALUE;
+                    long e = 0L;
+                    
+                    while (r != 0) {
+                        d = done;
+                        Object v = q.poll();
+                        empty = v == null;
+                        if (checkTerminated(d, empty, s)) {
+                            return;
+                        }
+                        if (empty) {
+                            break;
+                        }
+                        T value = nl.getValue(v);
+                        try {
+                            s.onNext(value);
+                        } catch (Throwable ex) {
+                            q.clear();
+                            Exceptions.throwIfFatal(ex);
+                            s.onError(OnErrorThrowable.addValueAsLastCause(ex, value));
+                            return;
+                        }
+                        r--;
+                        e++;
+                    }
+                    if (!unlimited && e != 0L) {
+                        addAndGet(-e);
+                    }
+                }
+                
+                synchronized (this) {
+                    if (!missed) {
```

Comment:
I see. You are saying "if there have not been additional requests while emitting then we should return". At first this seemed like a race condition however I see that the onNext would just hop threads to the other thread in that case. Do you think it would give us better performance and possible read easier by taking out the synchronized block and the `missed` field and replacing it with a check after each for loop iteration `if (get() > 0)`?